exporter/loadbalancingexporter/resolver_srv.go

```diff
@@ -0,0 +1,252 @@
+		target := strings.TrimSuffix(srv.Target, ".")
+		backendsWithIPs[target] = ""
+	}
+	// backends is what we use to compare against the current endpoints
+	var backends []string
+
+	// Lookup the IP addresses for the A records
+	for aRec := range backendsWithIPs {
+
+		// handle backends first
+		backend := aRec
+		// if a port is specified in the configuration, add it
+		if r.port != "" {
+			backend = fmt.Sprintf("%s:%s", backend, r.port)
+		}
+		backends = append(backends, backend)
+
+		ips, err := r.resolver.LookupIPAddr(ctx, aRec)
+		// Return the A record. If we can't resolve them, we'll try again next iteration
+		if err != nil {
+			_ = stats.RecordWithTags(ctx, srvResolverSuccessFalseMutators, mNumResolutions.M(1))
+			continue
+		}
+		// A headless Service SRV target only returns 1 IP address for its A record
+		if len(ips) > 1 {
+			return nil, errNotSingleIP
+		}
+
+		ip := ips[0]
+		if ip.IP.To4() != nil {
+			backendsWithIPs[aRec] = ip.String()
+		} else {
+			// it's an IPv6 address
+			backendsWithIPs[aRec] = fmt.Sprintf("[%s]", ip.String())
+		}
+	}
+
+	var freshBackends []string
+	for endpoint := range backendsWithIPs {
+		// If the old map doesn't have the endpoint, it's fresh
+		if _, ok := r.endpointsWithIPs[endpoint]; !ok {
+			freshBackends = append(freshBackends, endpoint)
+			// If the old map has the endpoint and IPs match it's still fresh
+			// Else freshBackends will be smaller and used later during callbacks
+		} else if backendsWithIPs[endpoint] == r.endpointsWithIPs[endpoint] {
+			freshBackends = append(freshBackends, endpoint)
+		}
+	}
+
+	// keep both in the same order
+	slices.Sort(freshBackends)
+	slices.Sort(backends)
+
+	if equalStringSlice(r.endpoints, freshBackends) {
+		r.logger.Debug("No change in endpoints")
+		return r.endpoints, nil
+	}
+
+	// the list has changed!
+	r.updateLock.Lock()
+	r.logger.Debug("Updating endpoints", zap.Strings("new endpoints", backends))
+	r.logger.Debug("Endpoints with IPs", zap.Any("old", r.endpointsWithIPs), zap.Any("new", backendsWithIPs))
+	r.endpoints = backends
+	r.endpointsWithIPs = backendsWithIPs
+	r.updateLock.Unlock()
+	_ = stats.RecordWithTags(ctx, srvResolverSuccessTrueMutators, mNumBackends.M(int64(len(backends))))
+
+	// propagate the change
+	r.changeCallbackLock.RLock()
+	for _, callback := range r.onChangeCallbacks {
+		// If backends != freshBackends it means an endpoint needs refreshed
+		if !equalStringSlice(r.endpoints, freshBackends) {
+			r.logger.Debug("Stale endpoints present", zap.Strings("fresh endpoints", freshBackends))
+			callback(freshBackends)
+		}
+		callback(r.endpoints)
+	}
+	r.changeCallbackLock.RUnlock()
+
+	return r.endpoints, nil
+}
+
+func (r *srvResolver) onChange(f func([]string)) {
+	r.changeCallbackLock.Lock()
+	defer r.changeCallbackLock.Unlock()
+	r.onChangeCallbacks = append(r.onChangeCallbacks, f)
+}
+
+func parseSRVHostname(srvHostname string) (service string, proto string, name string, err error) {
```

Comment:
instead of returning more than two parameters, it would be preferable to return a custom struct.